Instructor: edit question: do not warn about deleting responses if destructive changes are cancelled #7678

VamsiSangam opened the pull request:
Fixes #7678 

Apologies for the unusually delayed PR. I made some unnecessary changes which lead to chain of more complicated fixes, so I lost my way in the middle. I had to revert quite a lot of code. :sweat_smile: . Current implementation still has a few bugs. Will request for a review as soon as ready.

Status - ongoing.

VamsiSangam commented:
**Outline of Solution**

- Bug: Edit > Save contribution question without making any changes, results in delete responses. Solved it by removing a line which simulates a click, (click was causing the issue). [Diff](https://github.com/TEAMMATES/teammates/pull/7759/files#diff-5c827ae59bce58681fac5eddccc530cb).
- Bug: hidden inputs `showreceiverto`, `showresponsesto`, `showanswersto` of all questions were having the same value. Solved it by correcting a bug in jQuery selector string. [Diff](https://github.com/TEAMMATES/teammates/pull/7759/commits/99d3bb9c0fcec2621a15019a2362e292d9110d68).
- Destructive changes which cancel each other still trigger the delete responses modal: This includes changes like -
    - setting feedback path to a new value and reverting it.
    - changing visibility value and reverting it.
    - adding a rubric column/row and removing it.

There are many such above destructive changes which cancel each other. All-in-one solution for that -

- Revised `nonDestructive` fields. Marked helper `input` tags like counters as `nonDestructive`.
- Store the values of all form elements which don't have `.nonDestructive` class. Used jQuery's `clone` for this.
- While submitting the form, check these fields again. If -
    - The number of entries are the same, and
    - Each entry's value is still the same,
  then `mustDeleteResponses` attribute of `form` tag can be removed. This logic is executed only if the question has existing responses.

Apologies for the delayed PR. I had to revert many approaches :sweat_smile: 

Ready for review.

leeyimin commented on src/main/webapp/dev/js/main/instructorFeedbackEdit.es6:
```diff
@@ -611,6 +688,12 @@ function restoreOriginal(questionNum) {
         $(`#questionnum-${questionNum}`).val(questionNum);
         $(`#questionnum-${questionNum}`).prop('disabled', true);
+
+        const $form = $('#form_editquestion-1');
```
I believe you should be using `${questionNum}` here?

thenaesh commented on src/main/webapp/dev/js/main/instructorFeedbackEdit.es6:
```diff
@@ -387,6 +388,78 @@ function backupQuestion(questionNum) {
+
+    for (let i = 0; i < LENGTH; i += 1) {
+        const prevVal = $($prev.get(i)).val();
+        const currVal = $($curr.get(i)).val();
```
Is it possible to do this instead?

```javascript
        if (!$currVal.is($prevVal)) {
            return true;
        }
```

VamsiSangam replied:
I've tried using `.is()`, it doesn't seem to be working. Using `.val()` is giving the expected result.

thenaesh commented on src/main/webapp/dev/js/main/instructorFeedbackEdit.es6:
```diff
@@ -387,6 +388,78 @@ function backupQuestion(questionNum) {
+        return true;
+    }
+
+    const LENGTH = $prev.length;
```
If you want to do this, it should be lowercase, Maybe call it `numFields`.

VamsiSangam commented on src/main/resources/feedbackQuestionConstSumEditFormTemplate.html:
```diff
@@ -12,7 +12,7 @@
         </div>
 
-        <input type="hidden" name="${Const.ParamsNames.FEEDBACK_QUESTION_NUMBEROFCHOICECREATED}"
+        <input type="hidden" class="nonDestructive" name="${Const.ParamsNames.FEEDBACK_QUESTION_NUMBEROFCHOICECREATED}"
```
`nonDestructive` is actually a legacy class, I didn't create it. To change this, I'll have to look where else it is used. Shall I proceed with the change?

thenaesh replied:
Never mind, then. Not in this PR.

leeyimin commented:
@VamsiSangam I believe the implementation is still somewhat buggy. When changing radio buttons or checkboxes of constant sum question, or changing checkbox of team contribution question, no warning is issued before saving, which deletes away the responses.

VamsiSangam commented:
Thanks for pointing it out @leeyimin . Will fix it :smile:  I was actually planning to add a lot more code before requesting for review, but thanks for the heads up!

VamsiSangam commented:
I have changed the destructive changes backup mechanism. I added tests to cover some of the destructive changes. Will add the rest soon. Please review my code so far.

VamsiSangam commented:
@leeyimin @thenaesh reminder :)

VamsiSangam commented:
Completed test cases.

Ready for review.

damithc commented:
@thenaesh @leeyimin reminder

leeyimin commented:
@VamsiSangam Sorry for the delay. I am still in the midst of reviewing this PR but I seem to have found a bug. I am currently testing on Firefox v54.

Steps to reproduce the bug:
1. Edit a num-scale question.
2. Change one of the number fields.
3. Attempt to save and click cancel.
4. Discard changes.
5. Edit the question again and change one of the number fields again.
6. Attempt to save.

No warning seems to show up on the second attempt.

leeyimin commented:
![image](https://user-images.githubusercontent.com/6365864/29713285-736e61d8-89d0-11e7-8b84-ca2e0422fa0b.png)
Not sure if lines 413 - 419 in instructorFeedbackEdit.es6 are causing some errors. Sometimes, when i press edit, there is no response on the frontend.

VamsiSangam commented:
> Not sure if lines 413 - 419 in instructorFeedbackEdit.es6 are causing some errors. Sometimes, when i press edit, there is no response on the frontend.

It happens to me too. IFRC, the `nodeType` error from `tinymce.min.js` error pops up on all branches.

leeyimin commented:
@VamsiSangam On a side note, you can add `nonDestructive` to text question recommended length as well.

damithc commented:
I too encounter `click on edit but nothing happens` now and then, but I put it down to me clicking in the wrong place. Not sure if it is related.

VamsiSangam commented:
> No warning seems to show up on the second attempt.

In reply to the bug you pointed out in [this](https://github.com/TEAMMATES/teammates/pull/7759#issuecomment-324900892) comment. This occurs due to #7784 . When you discard changes, the `input` fields loose the `onchage` event handler which modifies the `editstatus` value to `mustDeleteResponses`. As you can see, the event handlers attached by JS after discarding are gone (there's one left because it is added by HTML not JS) -

![capture](https://user-images.githubusercontent.com/11662001/29713882-4e55e180-89be-11e7-89e8-5f259dd5630f.JPG)

Since, the `editstatus` attribute of `form` doesn't change to `mustDeleteResponses` when editing `input` after a discard, my logic doesn't even run.

> On a side note, you can add nonDestructive to text question recommended length as well.

Working on this.

leeyimin commented:
@VamsiSangam I see. I think this issue needs to fixed for your PR to be merged. How about altering your logic slightly to check against the destructive changes as long as responses exist?

VamsiSangam commented:
> How about altering your logic slightly to check against the destructive changes as long as responses exist?

Will do.

VamsiSangam commented:
@leeyimin added the suggested features. Ready for review.

leeyimin commented on src/main/webapp/dev/js/main/instructorFeedbackEdit.es6:
```diff
@@ -389,6 +390,74 @@ function backupQuestion(questionNum) {
+        // The changes cancelled out each other.
+        $(`#form_editquestion-${questionNum}`).attr('editstatus', 'hasResponses');
+    } else {
+        $(`#form_editquestion-${questionNum}`).attr('editstatus', 'mustDeleteResponses');
```
Since the mechanism does not make use of the `mustDeleteReponses` anymore, what do you think of removing it from the mechanism?

VamsiSangam replied:
If the `else` wasn't there, then in [this situation](https://github.com/TEAMMATES/teammates/pull/7759#issuecomment-324897305), the warning wouldn't show up because there are no JS event handlers which change `editstatus` to `mustDeleteResponses`. Due to the absence of JS event handlers `editstatus` would remain `hasResponses` and upon saving, the responses get deleted without a warning. In such a situation, the `else` block here, will add `mustDeleteResponses` if there's a destructive change, leading to a warning modal to show up. So this is required.

leeyimin replied:
Sorry, I kind of misunderstood how this method is currently used, but my understanding was that if this method return true/false, having the `mustDeleteResponses` status is not necessary. Another point was that having code setting `mustDeleteResponses` to `hasResponses` when resetting, turning into `mustDeleteResponses` when changes are applied to destructive fields are redundant. I thought it would be good to clean those parts up.

VamsiSangam replied:
Sorry @leeyimin , but I didn't understand what you are trying to say :sweat_smile:

leeyimin replied:
I felt that other parts of code dealing with `mustDeleteResponses` could be removed, eg in line 1165
```
$('form[id|=form_editquestion]').find(':input').not('.nonDestructive').not('.visibilityCheckbox')
            .change(...);
```
since either way before saving, a check against the backup will be preformed.

It seems like it is even possible to remove the use of `mustDeleteResponses` completely if this method returns a boolean value, which can then be used to decide if a warning modal should be shown.

VamsiSangam commented:
@leeyimin removed `mustDeleteResponses`.

Ready for review.

leeyimin commented:
@thenaesh Hope you can help to review the js code!

leeyimin commented on src/test/java/teammates/test/cases/browsertests/FeedbackConstSumOptionQuestionUiTest.java:
```diff
@@ -161,6 +166,127 @@ public void testAddQuestionAction() throws Exception {
+
+        ______TS("CONST SUM: testing deleting a const sum option");
+        // delete last option, must display modal
+        feedbackEditPage.reloadPage();
```
remove reloadPage?

VamsiSangam replied:
Reloading page is necessary to clear the status message. But I guess the redundant ones can be removed. Will do.

leeyimin commented on src/test/java/teammates/test/cases/browsertests/FeedbackConstSumRecipientQuestionUiTest.java:
```diff
@@ -113,6 +118,70 @@ public void testAddQuestionAction() throws Exception {
+        feedbackEditPage.verifyStatus(Const.StatusMessages.FEEDBACK_QUESTION_EDITED);
+        feedbackEditPage.isAlertClassEnabledForVisibilityOptions(1);
+
+        // change giver, must display modal
```
Missing `______TS()`?

VamsiSangam replied:
I was a little confused where to add `______TS()` and where to put a normal comment. Can you clarify when `______TS()` needs to be used?

leeyimin commented on src/test/java/teammates/test/cases/browsertests/FeedbackContributionQuestionUiTest.java:
```diff
@@ -123,6 +126,46 @@ public void testAddQuestionAction() throws Exception {
+        feedbackEditPage.clickEditQuestionButton(1);
+        feedbackEditPage.clickSaveExistingQuestionButton(1);
+        feedbackEditPage.verifyStatus(Const.StatusMessages.FEEDBACK_QUESTION_EDITED);
+        feedbackEditPage.isAlertClassEnabledForVisibilityOptions(1);
```
Is this necessary?

VamsiSangam replied:
I used `isAlertClassEnabledForVisibilityOptions` here to check that the responses are not actually deleted.

If you remember, @whipermr5 pointed out a bug that saving a contribution question just after clicking the editing button, was prompting the response deletion modal. I fixed it and this test is to cover that behavior.

VamsiSangam commented:
@leeyimin ready for review.

VamsiSangam commented:
@leeyimin @thenaesh please review.

damithc commented:
@VamsiSangam seems like this PR stalled due to lack of reviews. Sorry about that. Are you available to resume?

VamsiSangam commented:
Yes sir, I will fix the conflicts in a day or two.

wkurniawan07 commented:
Closing due to an inactivity for a very long period. If anyone is interested to continue the work done here, feel free to reopen/request for reopen.
